### ISSUE TITLE
"Other" variable assignment in ordinalbin and categoricalbin

### ISSUE THREAD

**Issue by jthrilly** [codaco/Architect#607]
User functionality request: When working with the ordinalbin/categorticalbin interfaces, users need the option to assign a value to an "other" option.

Users can already create an "other" option,  but this request is about being able to assign a _value_ to that option.

### Implementation

- New configuration section for these interfaces.
- "Create 'other' option?" checkbox. When selected, configuration section expands.
- Configuration is essentially a single "field configuration" screen from the name generator, and has two options: Question prompt, and variable for storing 'other' value (which will be a variable chooser filtered to variables of type text, which also allows creation). Once again, this is essentially the 'edit field' screen from the name generator form, so perhaps this could be reused?

### In NC
- Catbin and ordbin interfaces parse new config options, and create an additional "bin" as needed. 
- When a node is dropped into this bin, a form display is triggered (similar to the new node form), which displays the question prompt and assigns the variable.
- If the user cancels this form, the node is removed from the "other" category, and is added back into the bin to be dropped again.

**Comment by wwqrd** [comment]
What's the thinking behind the variable chooser? Just thinking about how this looks in the data:

We have a categorical variable, which contains numbers or strings, but now may also contain another variable, which is just a string? I'm not sure that that is meaningful?

**Comment by jthrilly** [comment]
It will be a separate variable entirely. Think of it like this:

- Categorical variable that the user names `participant_race`
- "other" variable, that the user creates using this feature, called `participant_race_other`

These are separate in the codebook. The values of categorical variables are effectively arrays of 'membership' of options. Options are themselves string values. What the 'other' variable is modelling is any number of additional option values.

**Comment by jthrilly** [comment]
To add more info, I imagine the protocol config to look something like this:

```
    {
      "id": "ordinalBin1",
      "type": "OrdinalBin",
      "label": "Contact Freq for close ties",
      "subject": {
        "entity": "node",
        "type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466"
      },
      "prompts": [
        {
          "id": "contactFreq",
          "text": "Contact frequency ordinal variable for close ties only",
          "variable": "1096204b-48fe-444c-b642-4ab211f7f57c",
          "otherVariable": "1106204b-48fe-444c-b642-4ab211f7f57d",
          "otherVariableLabel": "Other",
          "bucketSortOrder": [
            {
              "property": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
              "direction": "desc"
            }
          ],
          "binSortOrder": [
            {
              "property": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
              "direction": "desc"
            },
            {
              "property": "c5fee926-855d-4419-b5bb-54e89010cea6",
              "direction": "desc"
            }
          ]
        },
        {
          "id": "discussFreq",
          "text": "Discussion Frequency variable, with a much longer prompt than the previous prompt. Please also note that this prompt uses the color override to style these bins differently than the main prompt.",
          "variable": "23c5a7b9-b553-4ff1-b515-6a81999773d2",
          "color": "ord-color-seq-2"
        }
      ]
    },
```

**Comment by wwqrd** [comment]
Can you give an example of what that would look like in the network data?

**Comment by jthrilly** [comment]
In the network data it would be just as if the user had created a text variable on a form. No different. So you would have nodes that look like this:

``` JavaScript
const nodeWithOther = {
  uuid: "234235235",
  attributes: {
    participant_race: [],
    participant_race_other: 'other race'
  }
}

const normalNode = {
  uuid: "234235236",
  attributes: {
    participant_race: ['race1'],
    participant_race_other: null
  }
}
```

**Comment by wwqrd** [comment]
Is this going to be the definitive way to do this? Wondering why we need to specify a variable?:
```js
// in protocol...
{
  codebook: {
    'fad25cad': {
      'name': 'category',
      'type': 'categorical',
      'options': [
        { label: 'one', value: 1 },
        { label: 'two', value: 2 },
      ],
    }
  },
}

// in network...
{
  network: {
    nodes: [
      {
        id: 1,
        attributes: {
          category: 1,
        },
      },
      {
        id: 1,
        attributes: {
          category: 'my custom option',
        },
      }
    ],
  },
}
```

**Comment by jthrilly** [comment]
> Wondering why we need to specify a variable

I just think its cleaner to do it this way. I don't like the idea of having node variable values that don't correspond to the constraints of the codebook variable definition.

**Comment by jthrilly** [comment]
An idea I have been thinking about this morning for this is, what if the "other configuration section" simply reused the form section from the name generator? The major difference is that this would allow the user to optionally assign lots of variables in response to an "other" category. I think this could be quite powerful. It would also let us reuse existing components. 

In the example schema above, `otherVariable` would be replaced with something like `otherForm`, which would contain a form config in the existing schema format. 

What do you think?

**Comment by jthrilly** [comment]
Closed by #613
